### PR TITLE
Make test_subscriber_cpp always fail gracefully.

### LINKED ISSUE #360: Failing rclpy-to-rclcpp WString pub/sub over Connext on Windows
## Bug report

**Required Info:**

- Operating System:
  - Windows 10
- Installation type:
  - From source
- Version or commit hash:
  - Current master at f65f852432090d1f2431d0cf496cbf4e2006922a
- DDS implementation:
  - RTI Connext
- Client library (if applicable):
  - rclcpp & rclpy

#### Steps to reproduce issue
Run `rclpy`-to-`rclcpp` pub/sub tests using Connext on Windows:

```
colcon test --event-handlers=console_direct+ --packages-select test_communication --ctest-args " -R" " ^test_publisher_subscriber__rclpy__rclcpp__rmw_connext_cpp"
```

#### Expected behavior
All tests pass.

#### Actual behavior

Subscription to a WString publisher fails with:

```
18: [test_subscriber-24] r
18: INFO:test_subscriber-24-stderr:[test_subscriber-24] r
18: [test_subscriber-24] eceived message does not match any expected message
18: INFO:test_subscriber-24-stderr:[test_subscriber-24] eceived message does not match any expected message
18: INFO:test_publisher-23-stdout:[test_publisher-23]
18: [ERROR] [test_subscriber-24]: process has died [pid 3156, exit code 3221226505, cmd 'C:/Users/Administrator/nuclearsandwich_ws/build/test_communication/Release/test_subscriber_cpp.exe WStrings /test_time_21_23_20'].
18: ERROR:test_subscriber-24:process has died [pid 3156, exit code 3221226505, cmd 'C:/Users/Administrator/nuclearsandwich_ws/build/test_communication/Release/test_subscriber_cpp.exe WStrings /test_time_21_23_20'].
```

#### Additional information

This issue does not manifest itself on `rclcpp`-to-`rclcpp` nor `rclpy`-to-`rclpy` nor `rclcpp`-to-`rclpy` tests.

The process terminates almost silently. Looking at the `rclcpp` subscriber [implementation](https://github.com/ros2/system_tests/blob/master/test_communication/test/test_subscriber.cpp#L51), this suggests that a crash occurs during `rclcpp::shutdown`.

### PR DESCRIPTION
**Pull request by hidmic**
Fixes #360.

### CODE REVIEW

**Comment by hidmic** [comment]
Windows CI (building for Release and testing up to `test_communication`):

* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6848/)

File: test_communication/test/test_subscriber.cpp [hunk 1]
```diff
@@ -141,7 +139,14 @@ int main(int argc, char ** argv)
     rclcpp::shutdown();
     return 1;
   }
-  rclcpp::spin(node);
+
+  try {
+    rclcpp::spin(node);
+  } catch (const std::exception & e) {
+    std::cerr << e.what() << std::endl;
```
Comment (dirk-thomas) [inline test_communication/test/test_subscriber.cpp]: Please use `fprintf` here (see https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#c-specific).

File: test_communication/test/test_subscriber.cpp [hunk 2]
```diff
@@ -141,7 +139,14 @@ int main(int argc, char ** argv)
     rclcpp::shutdown();
     return 1;
   }
-  rclcpp::spin(node);
+
+  try {
+    rclcpp::spin(node);
+  } catch (const std::exception & e) {
```
Comment (clalancette) [inline test_communication/test/test_subscriber.cpp]: Just to make sure I understand correctly: if we throw a `std::runtime_error` in the callback above, we won't get the corruption, but we'll fail the test, correct?

Assuming that is the case, then don't we need to also look into the root of the test failure?

Comment (hidmic) [reply test_communication/test/test_subscriber.cpp]: Yeap, the root of the test failure is #359.

Comment (clalancette) [reply test_communication/test/test_subscriber.cpp]: Ah, I see, I didn't understand before.  OK, then this is fine with me with @dirk-thomas comment fixed.  Thanks for the clarification.

File: test_communication/test/test_subscriber.cpp (hunk 1 above)
Comment (hidmic) [reply test_communication/test/test_subscriber.cpp]: Uh, how's that `fprintf` prevents `stderr` interleaving?

Comment (dirk-thomas) [reply test_communication/test/test_subscriber.cpp]: Both operators are executed separately and have a chance of being interleaved by any other concurrent write to `stdout`. The `printf` creates a single string including the newline and writes them in a single call to `stdout`. While the latter technicall doesn't guarantee the operation to be atomic it practically makes it much much less likely to be interleaved.

Comment (hidmic) [reply test_communication/test/test_subscriber.cpp]: Thanks for the explanation, fixed in bbc82e9.

**Comment by hidmic** [comment]
May I get an approval here? As to why CI is green, there's an issue with launch testing and exit codes on Windows that I need to fix. Long story short, I made the same mistake than the old launch testing did and process that exit with a 1 are taken as successful.

**Pull request merged**